### DISCUSSION TITLE
irregular application crash

### DISCUSSION THREAD

**Discussion by fernando-galaxystaking** [crypto-org-chain/chain-main#707] [Q&A] [answered]
Our application (v 3.3.3) has now crashed several times on different hardware with the following error message you see in attach.

Due to the automatic restart of the service, the application comes back, but in most cases it crashes again immediately or reports a db-error.
I then only get the node to run by importing a statesync db and sync the delta.

What means the error message? Where does this error message come from? Disk, RAM, CPU, Connection?
[error.log](https://github.com/crypto-org-chain/chain-main/files/8123117/error.log)

**Comment by adu-crypto** [comment, marked as answer]
it seems the type assertion here has failed and trigger a panic:
https://github.com/cosmos/cosmos-sdk/blob/master/server/logger.go#L51

**Reply by fernando-galaxystaking** [reply]
Thank you for your reply. 
So i can understand: How do you get this? By which Information in the error log?
What could be causing this? Any Idea?

**Reply by tomtau** [reply]
@alexanderbez mentioned on Discord
> the keys you provide to a logger must be a string
> e.g. logger.Debug("key1", 4, "key2", false)
> this tells me you have log that is prob something like this:
> `logger.Debug(4, 4)`

There isn't much custom logging in chain-main, so it looks like it is caused by logging inside Cosmos SDK and it could be caused by any logging keys that were passed in from a configuration, environment variables or command line arguments to chain-main.

**Reply by adu-crypto** [reply]
> Feb 22 22:13:38 node03 chain-maind[5681]: runtime: unexpected return pc for github.com/cosmos/cosmos-sdk/server.ZeroLogWrapper.Debug called from 0xc002462728

this indicates that there was an unexpected panic when calling `server.ZeroLogWrapper.Debug`

> node03 chain-maind[5681]: github.com/cosmos/cosmos-sdk/server.getLogFields(...)
Feb 22 22:13:38 node03 chain-maind[5681]:         /go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.44.5/server/logger.go:51

this indicates that the panic originated from `server.getLogFields`:

look into `server.getLogFields`:
```go
func getLogFields(keyVals ...interface{}) map[string]interface{} {
	if len(keyVals)%2 != 0 {
		return nil
	}

	fields := make(map[string]interface{})
	for i := 0; i < len(keyVals); i += 2 {
		fields[keyVals[i].(string)] = keyVals[i+1]
	}

	return fields
}
```
and there is a `runtime.panicdottypeE` in errorlog, so the failed type conversion should have triggered the panic.

This means the caller of this function has passed in not-string types as keys, so I try to find the caller.

> Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08c88:  0x000000c00b125d40  0x0000000000000000
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08c98:  0x000000c004a08d58  0x000000000043b894 <runtime.gopanic+0x0000000000000114>
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08ca8:  0x0000000000000000  0x000000c004a08d30
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08cb8:  0x0000000000413814 <runtime.mallocgc+0x00000000000006b4>  0x0000000000000000
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08cc8:  0x0000000000000030  0x00000000010edde7 <github.com/cosmos/cosmos-sdk/server.(*ZeroLogWrapper).Debug+0x00000000000000a7>
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08cd8:  0x0000000002420100  0x010000c00145b5f0
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08ce8:  0x00007f28c4709568  0x0000000000000020
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08cf8:  0x0000000000000028  0x00007f292f39fb38
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08d08:  0x0000000000000030  0x000000c00b125d60
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08d18:  0x000000c00b125d40  0x0000000000000000
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08d28:  0x00000000018ca640  0x000000c026ac2db0
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08d38:  0x0000000000000000  0x0000000000000000
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08d48:  0x0000000000000000  0x0000000000000000
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08d58:  0x000000c004a08d78  0x0000000000410ef5 <runtime.panicdottypeE+0x0000000000000075>
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08d68:  0x00000000018ca640  0x000000c026ac2db0
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08d78:  0x000000c004a08dc0  0x00000000010e7735 <github.com/cosmos/cosmos-sdk/server.ZeroLogWrapper.Debug+0x0000000000000235>
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08d88: <0x0000000000000000  0x000000000182ece0
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08d98:  0x000000000187c580  0x000000c00145b600
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08da8:  0x0000000000000001  0x0000000000000001
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08db8:  0x000000c004a08e07  0x000000c004a08e38
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08dc8:  0x0000000000c3bc90 <github.com/tendermint/tendermint/p2p/conn.(*MConnection).Send+0x00000000000001f0>  0x00000000018a92a0
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08dd8:  0x000000c0332cb350  0x000000c004a08e28
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08de8:  0x000000c0146e0660  0x0000000000000001
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08df8:  0x00000000004535d2 <runtime.makeslice+0x0000000000000052>  0x0100000000005517
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08e08:  0x3800000000000000  0x00000000c004a08e
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08e18:  0x000000c0146e0660  0x000000c013892000
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08e28:  0x0000000001810b00  0x000000c011227320
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08e38:  0x000000c004a08ea8  0x0000000000c503d3 <github.com/tendermint/tendermint/p2p.(*peer).Send+0x0000000000000093>
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08e48:  0x000000c01b2ec2c0  0x000000c00145b521
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08e58:  0x000000c01289a000  0x0000000000005517
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08e68:  0x0000000000005517  0x00007f290438e430
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08e78:  0x000000c02c4b5fa0  0x00000000000001f4
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08e88:  0x000000c004a08ea8  0x0000000000f8435d <github.com/tendermint/tendermint/consensus.MustEncode+0x000000000000003d>
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08e98:  0x000000000242da20  0x000000c02c4b5fa0
Feb 22 22:13:38 node03 chain-maind[5681]: 0x000000c004a08ea8:  0x000000c004a08fa0  0x0000000000f88b4d <github.com/tendermint/tendermint/consensus.(*Reactor).gossipDataRoutine+0x0000000000000a0d>

I follow the stack trace and look into `github.com/tendermint/tendermint/p2p/conn.(*MConnection).Send` which calls `server.ZeroLogWrapper.Debug`:
```go
func (c *MConnection) Send(chID byte, msgBytes []byte) bool {
	if !c.IsRunning() {
		return false
	}

	c.Logger.Debug("Send", "channel", chID, "conn", c, "msgBytes", fmt.Sprintf("%X", msgBytes))

	// Send message to channel.
	channel, ok := c.channelsIdx[chID]
	if !ok {
		c.Logger.Error(fmt.Sprintf("Cannot send bytes, unknown channel %X", chID))
		return false
	}

	success := channel.sendBytes(msgBytes)
	if success {
		// Wake up sendRoutine if necessary
		select {
		case c.send <- struct{}{}:
		default:
		}
	} else {
		c.Logger.Debug("Send failed", "channel", chID, "conn", c, "msgBytes", fmt.Sprintf("%X", msgBytes))
	}
	return success
}
```
so one of the `c.Logger.Debug` here should have triggered the panic
but arguments here seem alright.
At last, I think the failed type assertion has triggered this panic but I don't know why.
maybe the memory has corrupted?

**Reply by fernando-galaxystaking** [reply]
Thank you for the detailed explanation. That gives me a better understanding of the process.
Theoretically, if it really is the RAM, it should be possible to solve this with ECC-RAM, right? That would be an option for us.

Anyone have another theory/suggestion?

**Reply by alexanderbez** [reply]
Yeah something isn't adding up here. The stacktrace seems to indicate something else. It does seem like some sort of memory corruption or just a memory related issue. Does this happen consistently? Or did you only see this issue once?

**Reply by fernando-galaxystaking** [reply]
Litte Background:
We have several (>10) nodes / servers in use to map our infrastructure, since we have a total of 3 validators, 6 sentrys, Web-Server (Homepage), 2 backup servers with running chain, statesync server to make snapshots (https://statesync.galaxystaking.space) and watchdog for failure.

This problem has now occurred about 4-5 times on different servers.
However, as a precaution, they were completely replaced by the hoster, but problem still exists sporadically.

In the next Step i would let replace the RAM with ECC-RAM. This would solve the problem right?

**Reply by alexanderbez** [reply]
Is it deterministic? Like does it always happen at a specific block height? Or is it somewhat random?

**Reply by fernando-galaxystaking** [reply]
Unfortunately not predictable. It's always on a different block.
In the event of a failure, our watchdog kicked in, the validator is pushed to a backup server and continues to run there.
The "defective" one is then loaded with a fresh statesync DB (to save space and rule out possible DB errors) and is available again as a backup.
It is quite possible that without a DB error, the chain will start up again after a few minutes and continue running. But I won't let it get that far, because the number of missed blocks is important.
Next time (if comes :D) i will use / check the old DB after moved to backup server.

**Reply by fernando-galaxystaking** [reply]
I'm currently still wondering whether ECC would solve this problem?
